hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,525 @@
+            Some(Proof(merkle_proof)) => Ok(IndexedMerkleProof {
+                index,
+                count,
+                merkle_proof,
+            }),
+        }
+    }
+
+    #[allow(unused)]
+    pub(crate) fn root_hash(&self) -> Digest {
+        use blake2::{
+            digest::{Update, VariableOutput},
+            VarBlake2b,
+        };
+
+        let IndexedMerkleProof {
+            index: _,
+            count,
+            merkle_proof,
+        } = self;
+
+        let mut hashes = merkle_proof.iter();
+        let raw_root = if let Some(leaf_hash) = hashes.next().cloned() {
+            // Compute whether to hash left or right for the elements of the Merkle proof.
+            // This gives a path to the value with the specified index.
+            // We represent this path as a sequence of 64 bits. 1 here means "hash right".
+            let mut path: u64 = 0;
+            let mut n = self.count;
+            let mut i = self.index;
+            while n > 1 {
+                path <<= 1;
+                let pivot = 1u64 << (63 - (n - 1).leading_zeros());
+                if i < pivot {
+                    n = pivot;
+                } else {
+                    path |= 1;
+                    n -= pivot;
+                    i -= pivot;
+                }
+            }
+
+            // Compute the raw Merkle root by hashing the proof from leaf hash up.
+            let mut acc = leaf_hash;
+
+            for hash in hashes {
+                let mut hasher = VarBlake2b::new(Digest::LENGTH).unwrap();
+                if (path & 1) == 1 {
+                    hasher.update(hash);
+                    hasher.update(&acc);
+                } else {
+                    hasher.update(&acc);
+                    hasher.update(hash);
+                }
+                hasher.finalize_variable(|slice| {
+                    acc.0.copy_from_slice(slice);
+                });
+                path >>= 1;
+            }
+            acc
+        } else {
+            Digest::SENTINEL_MERKLE_TREE
+        };
+
+        // The Merkle root is the hash of the count with the raw root.
+        Digest::hash_pair(count.to_le_bytes(), raw_root)
+    }
+
+    pub(crate) fn merkle_proof(&self) -> &[Digest] {
+        &self.merkle_proof
+    }
+
+    // Proof lengths are never bigger than 65, so we can use a u8 here
+    // The reason they are never bigger than 65 is because we are using 64 bit counts
+    pub(crate) fn compute_expected_proof_length(&self) -> u64 {
+        if self.count == 0 {
+            return 0;
+        }
+        let mut l = 1;
+        let mut n = self.count;
+        let mut i = self.index;
+        while n > 1 {
+            let pivot = 1u64 << (63 - (n - 1).leading_zeros());
+            if i < pivot {
+                n = pivot;
+            } else {
+                n -= pivot;
+                i -= pivot;
+            }
+            l += 1;
+        }
+        l
+    }
+
+    #[allow(unused)]
```

Comment:
As above, why allow unused?

Reply:
Initially it was `#[cfg(test)]`, but there was a discussion indicating that we should favor `#[allow(unused)]` for such cases.
Especially in this case, because `verify()` will be used by code outside of the `tests` module in the future.

But, anyway - thanks to the other comment this function is now used, therefore I removed the attribute at all.